dbt/models/reporting/reporting.vw_assessment_progress.sql

```diff
@@ -0,0 +1,172 @@
+        ARRAY_JOIN(vpu.tax_school_unified_district_name, ', ')
+            AS school_unified_district_name,
+        CASE
+            WHEN
+                vpu.ward_chicago_data_year IS NOT NULL
+                THEN REPLACE(vpu.ward_name, 'chicago_', '')
+        END
+            AS chicago_ward_name
+    FROM {{ ref('default.vw_pin_universe') }} AS vpu
+    LEFT JOIN
+        (
+            SELECT DISTINCT
+                pin,
+                year
+            FROM {{ ref('reporting.vw_pin_value_long') }}
+        )
+            AS pins
+        ON vpu.pin = pins.pin
+        AND vpu.year = pins.year
+    WHERE pins.pin IS NOT NULL
+        OR vpu.year
+        = (SELECT MAX(year) FROM {{ ref('default.vw_pin_universe') }})
+
+),
+
+-- Cross join our trimmed sample with the stages we defined above
+expanded_geos1 AS (
+    SELECT
+        trimmed_geos.*,
+        stages.*
+    FROM trimmed_geos
+    CROSS JOIN stages
+),
+
+-- Join on values by stage from vw_pin_value_long so we can determine which
+-- parcels have values for which stages in which years.
+expanded_geos2 AS (
+    SELECT
+        expanded_geos1.*,
+        CASE WHEN vpvl.tot IS NOT NULL THEN 1 ELSE 0 END AS has_value,
+        vpvl.bldg,
+        vpvl.land,
+        vpvl.tot
+    FROM expanded_geos1
+    LEFT JOIN {{ ref('reporting.vw_pin_value_long') }} AS vpvl
+        ON expanded_geos1.pin = vpvl.pin
+        AND expanded_geos1.year = vpvl.year
+        AND expanded_geos1.stage_name = vpvl.stage_name
+),
+
+/* Calculate the denominator for the pct_pin_w_value_in_group column.
+default.vw_pin_universe serves as the universe of yearly PINs we expect
+to see in reporting.vw_pin_value_long. Loop over the different geographies
+we track progress for and then stack the output.*/
+pin_counts AS (
+
+    {{ assessment_progress_pin_count(
+            from = 'expanded_geos2',
+            geo_type = 'Municipality',
+            column_name = 'municipality_name'
+            ) }}
+    UNION ALL
+    {{ assessment_progress_pin_count(
+            from = 'expanded_geos2',
+            geo_type = 'Chicago Ward',
+            column_name = 'chicago_ward_name'
+            ) }}
+    UNION ALL
+    {{ assessment_progress_pin_count(
+            from = 'expanded_geos2',
+            geo_type = 'Elementary School District',
+            column_name = 'school_elementary_district_name'
+            ) }}
+    UNION ALL
+    {{ assessment_progress_pin_count(
+            from = 'expanded_geos2',
+            geo_type = 'Secondary School District',
+            column_name = 'school_secondary_district_name'
+            ) }}
+    UNION ALL
+    {{ assessment_progress_pin_count(
+            from = 'expanded_geos2',
+            geo_type = 'Unified School District',
+            column_name = 'school_unified_district_name'
+            ) }}
+)
+
+-- Calculate the portion of each geography that has progressed through an
+-- assessment stage by class.
+SELECT
```

Comment:
**thought (non-blocking):** This is a pretty wild query. It took me a bit of running the CTEs to understand exactly what's happening. Now that we've moved fully onto dbt, it might be worth decomposing the CTEs into individual models or ephemeral views (once they're ready). The goal is to enable folks without your domain knowledge to work on/debug parts of these larger views.

An alternative to decomposing: even though the commenting here is really good, it might be clearer/even more helpful to add ASCII table representations of the data at each step/CTE.

Reply:
Yeah, I agree this query is a little rough on the brain. Between it's complexity and the amount of data it queries, I'll try to sketch out an alternative this week.